### PR TITLE
After Split_part function fix

### PR DESCRIPTION
**Pull request by SaiTejaVenkatapuram**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 1]
```diff
@@ -1396,6 +1399,31 @@ protected void unparseDateTime(SqlWriter writer, SqlCall call, int leftPrec, int
...
   }
 
+  private void unparseSplitPartFunction(SqlWriter writer, SqlCall call,
+      int leftPrec, int rightPrec) {
+    SqlWriter.Frame splitFrame = writer.startFunCall("SPLIT");
+    call.operand(0).unparse(writer, leftPrec, rightPrec);
+    writer.print(",");
+    call.operand(1).unparse(writer, leftPrec, rightPrec);
+    writer.endFunCall(splitFrame);
+    writer.print("[SAFE_OFFSET (");
```
Comment (Katakam-Phani-Mahidhar) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:1409]: refer SqlStdOperatorTable.ITEM for implementation of array access

File: core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java [hunk 2]
```diff
@@ -1396,6 +1399,31 @@ protected void unparseDateTime(SqlWriter writer, SqlCall call, int leftPrec, int
...
+    SqlNode thirdOperandValue = call.operand(2);
+    if (thirdOperandValue instanceof SqlLiteral) {
+      SqlLiteral thirdOperandLiteral = (SqlLiteral) thirdOperandValue;
+      if (thirdOperandLiteral.getValue() == null) {
+        writer.print(null);
+      } else {
+        int thirdOperand = Integer.valueOf(call.operand(2).toString());
+        if (thirdOperand == 0) {
```
Comment (Katakam-Phani-Mahidhar) [inline core/src/main/java/org/apache/calcite/sql/dialect/BigQuerySqlDialect.java:1417]: pls move this to BQConverterRexShuttle

**Review by Katakam-Phani-Mahidhar — CHANGES_REQUESTED** [review]
Please make another operator in SqlLibraryOperators for safe array access in big query analogous to SqlStdOperatorTable.ITEM

**Review by nosheenSheikh02 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 3]
```diff
@@ -1714,4 +1742,14 @@ private SqlLibraryOperators() {
...
       new SqlMedianAggFunction(SqlKind.MEDIAN, ReturnTypes.ARG0_NULLABLE);
 
+  @LibraryOperator(libraries = {SNOWFLAKE, BIG_QUERY})
+  public static final SqlFunction SPLIT_PART = new SqlFunction(
+      "SPLIT_PART",
+      SqlKind.OTHER_FUNCTION,
+      ReturnTypes.VARCHAR_2000_NULLABLE,
+      null,
+      OperandTypes.or(OperandTypes.STRING_STRING_INTEGER,
+          OperandTypes.family(SqlTypeFamily.NULL, SqlTypeFamily.STRING, SqlTypeFamily.INTEGER)),
```
[the remark this answers is not in the document]
Comment (SaiTejaVenkatapuram) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: OperandTypes.family(SqlTypeFamily.NULL, SqlTypeFamily.STRING, SqlTypeFamily.INTEGER) changed to OperandTypes.NULL_STRING_INTEGER

File: core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java [hunk 4]
```diff
@@ -311,6 +311,11 @@ public static SqlOperandTypeChecker variadic(
   public static final SqlSingleOperandTypeChecker ARRAY =
       family(SqlTypeFamily.ARRAY);
 
+  public static final SqlSingleOperandTypeChecker ARRAY_OR_MAP =
+      OperandTypes.family(SqlTypeFamily.ARRAY)
+          .or(OperandTypes.family(SqlTypeFamily.MAP))
+          .or(OperandTypes.family(SqlTypeFamily.ANY));
+
```
[the remark this answers is not in the document]
Comment (SaiTejaVenkatapuram) [reply core/src/main/java/org/apache/calcite/sql/type/OperandTypes.java:318]: Made changes as mentioned

File: core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java [hunk 5]
```diff
@@ -107,36 +103,47 @@ public SqlItemOperator(boolean isInvalidIndexAllowed) {
...
           requireNonNull(sqlTypeName.getFamily(),
               () -> "keyType.getSqlTypeName().getFamily() null, type is " + sqlTypeName));
     case ROW:
     case ANY:
     case DYNAMIC_STAR:
-      return OperandTypes.or(
-          OperandTypes.family(SqlTypeFamily.INTEGER),
-          OperandTypes.family(SqlTypeFamily.CHARACTER));
+      return OperandTypes.family(SqlTypeFamily.INTEGER)
+          .or(OperandTypes.family(SqlTypeFamily.CHARACTER));
```
[the remark this answers is not in the document]
Comment (SaiTejaVenkatapuram) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlItemOperator.java]: Reverted back the changes

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 6]
```diff
@@ -38,10 +38,13 @@
 import org.apache.calcite.sql.type.SqlTypeFamily;
 import org.apache.calcite.sql.type.SqlTypeName;
 import org.apache.calcite.sql.type.SqlTypeTransforms;
+import org.apache.calcite.sql.type.SqlTypeUtil;
```
[the remark this answers is not in the document]
Comment (SaiTejaVenkatapuram) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Removed unused imports

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 7]
```diff
@@ -38,10 +38,13 @@
...
 import org.apache.calcite.util.Optionality;
 
 import com.google.common.collect.ImmutableList;
 
+import org.apache.calcite.util.Static;
+
```
[the remark this answers is not in the document]
Comment (SaiTejaVenkatapuram) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: Removed unused imports

File: core/src/main/java/org/apache/calcite/sql/type/SqlSingleOperandTypeChecker.java [hunk 8]
```diff
@@ -68,4 +68,16 @@ boolean checkSingleOperandType(
...
+  default SqlSingleOperandTypeChecker and(
+      SqlSingleOperandTypeChecker checker) {
+    return OperandTypes.and(this, checker);
+  }
+
+  /** Composes this with another single-operand checker using OR. */
+  default SqlSingleOperandTypeChecker or(
+      SqlSingleOperandTypeChecker checker) {
+    return OperandTypes.or(this, checker);
+  }
```
[the remark this answers is not in the document]
Comment (SaiTejaVenkatapuram) [reply core/src/main/java/org/apache/calcite/sql/type/SqlSingleOperandTypeChecker.java]: Removed the changes.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 9]
```diff
@@ -1115,11 +1142,12 @@ private SqlLibraryOperators() {
...
       SqlFunctionCategory.STRING);
 
-  @LibraryOperator(libraries = {HIVE, SPARK})
+  @LibraryOperator(libraries = {HIVE, SPARK, BIG_QUERY})
   public static final SqlFunction SPLIT = new SqlFunction(
       "SPLIT",
       SqlKind.OTHER_FUNCTION,
-      ReturnTypes.MULTISET_NULLABLE,
+      ReturnTypes.ARG0
+          .andThen(SqlTypeTransforms.TO_ARRAY),
```
[the remark this answers is not in the document]
Comment (Katakam-Phani-Mahidhar) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:1157]: The current return type derivation does not work when string type is added as an operand. So, this is added in order to get the current ReturnType.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java [hunk 10]
```diff
@@ -12808,6 +12809,41 @@ private String toSqlWithBloat(RelNode root, int bloat) {
...
+        .scan("EMP")
+        .project(builder.alias(splitAccess, "Result"))
+        .build();
+
+    final String expectedBigQuery = "SELECT SPLIT('123@Domain|Example', '@')[SAFE_OFFSET(2)] "
+        + "AS Result\nFROM scott.EMP";
+
+    assertThat(toSql(root, DatabaseProduct.BIG_QUERY.getDialect()),
+        isLinux(expectedBigQuery));
+  }
```
[the remark this answers is not in the document]
Comment (Katakam-Phani-Mahidhar) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:12845]: As the SPLIT operator ReturnType is modified. this test case is added to test that change.

**Review by nosheenSheikh02 — APPROVED** [review]

**Review by dm-shantanuc — APPROVED** [review]
LGTM. Also reviewed by Anurag.

**Review by dm-shantanuc — APPROVED** [review]

**Review by Katakam-Phani-Mahidhar — APPROVED** [review]

**Pull request merged**